cxx-qt-build: Add more information about qmake, how it is used and why (#497)
KDAB/cxx-qt: book/src/getting-started/4-cmake-integration.md

- Add `cxx`, `cxx-qt`, as well as `cxx-qt-lib` as dependencies
- Add `cxx-qt-build` as a build dependency

Note that in order for cxx-qt to work, the `qmake` executable must be located. This is because cxx-qt relies on `qmake` to locate the necessary Qt libraries and header files on your system. This can be done by either adding it to your system's `PATH`, or by setting the `QMAKE` environment variable to its path.

> To check which version of Qt will be used with `qmake`, you can use the `qmake -query` command. This will display information about the Qt installation, including the version number and installation path.

Reviewer: This is helpful :+1:

Reviewer: > To check which version of Qt will be used with `qmake`, you can use the `qmake -query` command. This will display information about the Qt installation, including the version number and installation path.

This isn't always correct. CMake can pick up a different version of Qt than the first `qmake` executable found in `PATH`, particularly if the user sets `CMAKE_PREFIX_PATH`.

Author: But we are in the cargo/build.rs section, so everything is based on the qmake exe right?

Of course from the CMake, the `find_package` as so many option and could behave in some many way. But this is not what this section is about.

Reviewer 2: With CMake, the location that CMake chooses for `QMAKE` is passed down into Rust, so if it picked one outside of your `PATH` this should be passed into the Rust env to IIRC

I think this is done via these places
https://github.com/KDAB/cxx-qt/blob/c435d65aba8eef451596def5f29c3b52dae8ea23/examples/qml_minimal/CMakeLists.txt#L29

 https://github.com/KDAB/cxx-qt/blob/c435d65aba8eef451596def5f29c3b52dae8ea23/examples/qml_minimal/CMakeLists.txt#L51-L60

Author: So I think this documentation should move to the the cargo build system page.
On the cargo section in the cmake page, I shoudn't mention the PATH at all. Just QMAKE, and say that it will later be passed from whatever qt version CMake found as an env variable. Would it be ok?

Author: I haven't look in cargo only build yet. When you say you don't have to ahve `QMAKE` set, you still need to have it in the `PATH` right?

Reviewer 2: I think that's how it's working there yes, but i'd have to look at the code again.

The `QMAKE` env var is for ensuring that CMake and Cargo when combined agree on the same `qmake` executable.

Author: Ok so on rust side it is check for `QMAKE` env variable, then fallback to `qmake` in `PATH`, then error.
In the CMake documentation, only the `QMAKE` is interesting to the user.

Reviewer: Correct. The reason we must pass the QMAKE environment variable in CMake builds is to ensure that whatever Qt installation CMake picks up due to CMAKE_PREFIX_PATH or whatever other mechanism is the same Qt installation that build.rs uses.

Every Rust source file that uses the `#[cxx_qt::bridge]` macro need to be included in this script.
In our case, this is only the `src/cxxqt_object.rs` file.

> > At this point, you can test your library build by running `QMAKE=/path/to/qmake cargo build` inside the `rust` folder. This should output a static library in `rust/target/debug/qml_minimal.lib` on Windows or `rust/target/debug/qml_minimal.a` on Unix system.

Reviewer: Do we really need to say this? I'd like to keep the documentation concise.

Author: Maybe the second sentence is less useful, but I really think the first one really is.
When you think about it the whole setup is quite complex as it require understanding of cpp, rust, cargo, cmake, qml, qrc, qt. Being able to test at this without bringing further complication is very helpful. 
Having this note here is useful to debug the cargo part of the build without bringing cmake in, finding qt from cmake issues etc...
I aware than when has been done once it is trivial, but on first run, I was happy to just test that part. It worked out of the box my system because I had qmake in the path, I think I would have struggle a bit more or I wouldn't understand everything if I had just runner everything from cmake.

Reviewer: I think this suggestion could open a can of worms for users who may think they need to run Cargo separately from CMake. Users might also get confused if running `cargo build` picks up a different version of qmake than CMake finds. If someone wants to play around with Cargo without involving CMake at all, they can run `cargo build -p qml-minimal-no-cmake` as explained in the following chapter.

Author: Ok, I haven't read this chapter yet, as I'm interesting in integrating cxx-qt in already existing CMake based application.
I will remove this line because I see your point.

I don't know if people will behave like, but I think they will go either to the building with cmake or cargo page. In the cargo page there is no mention at all of qmake.

Maybe I could duplicate the information about qmake to that other page? But I guess that would be another PR

{{#include ../../../examples/qml_minimal/CMakeLists.txt:book_cmake_setup}}
```

> To ensure that cxx-qt-build uses the same version of Qt as your CMake targets, you can use the `Qt` CMake target to locate the qmake executable. The `qmake` executable will be pass to `build.rs` as an environment variable using `corrosion_set_env_vars`.

Reviewer: ```suggestion
To ensure that cxx-qt-build uses the same version of Qt as your CMake targets, use the `Qt` CMake target to locate the qmake executable. Then, the pass `qmake` executable path to `build.rs` with an environment variable using `corrosion_set_env_vars`.
```
